### PR TITLE
Subfeature/serialization framework

### PR DESCRIPTION
**Pull request by GlynLeine**

### CODE REVIEW

File: applications/sandbox/systems/examplesystem.hpp [hunk 1]
```diff
@@ -4,15 +4,61 @@
...
+            member_reference
+            {
+                "value",
+                primitive_reference{ typeHash<int>(), &obj.value }
+            }
+        };
+        refl.data = reinterpret_cast<void*>(adress);
+        return refl;
+    }
+}
```
Comment (GlynLeine) [inline applications/sandbox/systems/examplesystem.hpp:58]: this is temporary until the pre-processor is merged into this branch

File: legion/engine/core/types/meta.hpp [hunk 2]
```diff
@@ -13,7 +13,6 @@
 namespace legion::core
 {
 
-#if !defined(DOXY_EXCLUDE)
 #define HAS_FUNC(x)                                                                                                     \
```
Comment (GlynLeine) [inline legion/engine/core/types/meta.hpp]: mistake with rebase, needs to be undone

File: legion/engine/core/types/meta.hpp [hunk 3]
```diff
@@ -61,25 +60,7 @@ namespace legion::core
...
-    constexpr bool CONCAT(has_, CONCAT(x, _v)) = CONCAT(has_, x)<C, F>::value;                                   \
-                                                                                                                        \
-    template<typename, typename T>                                                                                      \
-    struct CONCAT(has_static_, x) {                                                                                     \
-        static constexpr bool value = false;                                                                            \
-    };                                                                                                                  \
-                                                                                                                        \
-    template<typename C, typename F>                                                                                    \
-    constexpr bool CONCAT(has_static_, CONCAT(x, _v)) = CONCAT(has_static_, x)<C, F>::value;
-#endif
```
Comment (GlynLeine) [inline legion/engine/core/types/meta.hpp]: mistake with rebase, needs to be undone

File: legion/engine/core/types/meta.hpp [hunk 4]
```diff
@@ -103,39 +84,69 @@ namespace legion::core
...
+        ->std::false_type;                                                                                              \
                                                                                                                         \
-        typedef decltype(check<EXPAND(CAT_PREFIX(_, templateArgs))>(nullptr)) type;                                     \
+      typedef decltype(check<void>(nullptr)) type;                                                                      \
     public:                                                                                                             \
-        static constexpr bool value = type::value;                                                                      \
+      static constexpr bool value = type::value;                                                                        \
     };                                                                                                                  \
-                                                                                                                        \
```
Comment (GlynLeine) [inline legion/engine/core/types/meta.hpp]: mistake with rebase, needs to be undone

File: legion/engine/core/types/meta.hpp [hunk 5]
```diff
@@ -162,13 +173,10 @@ namespace legion::core
...
 
     template<template<typename...>typename T, typename U, size_type I, typename... Args>
-    struct make_sequence
-#if !defined(DOXY_EXCLUDE)
-        : make_sequence<T, U, I - 1, Args..., U>{};
+    struct make_sequence : make_sequence<T, U, I - 1, Args..., U> {};
 
     template<template<typename...>typename T, typename U, typename... Args>
     struct make_sequence<T, U, 0, Args...>
-#endif
```
Comment (GlynLeine) [inline legion/engine/core/types/meta.hpp]: mistake with rebase, needs to be undone

File: legion/engine/core/types/meta.hpp [hunk 6]
```diff
@@ -187,40 +195,38 @@ namespace legion::core
...
-    struct element_at
-#if !defined(DOXY_EXCLUDE)
-        : element_at<I - 1, Types...>
+    struct element_at : element_at<I - 1, Types...>
     {
     };
 
     template<typename Type, typename... Types>
     struct element_at<0, Type, Types...>
-#endif
```
Comment (GlynLeine) [inline legion/engine/core/types/meta.hpp]: mistake with rebase, needs to be undone

File: legion/engine/core/types/meta.hpp [hunk 7]
```diff
@@ -187,40 +195,38 @@ namespace legion::core
...
-        template <typename...>
-        static constexpr auto check(...)
-            ->std::false_type;
+    template<class T, typename... Args>
+    std::false_type
+        brace_construct_test(...);
 
-        using type = decltype(check<T, Args...>(nullptr));
-    public:
-        static constexpr bool value = type::value;
```
Comment (GlynLeine) [inline legion/engine/core/types/meta.hpp]: mistake with rebase, needs to be undone

File: legion/engine/core/ecs/registry.inl [hunk 8]
```diff
@@ -61,6 +61,7 @@ namespace legion::core::ecs
             {
                 checked = true;
                 registerComponentType<component_type>();
+                //serialization::serializer_registry::register_serializer<component_type>();
```
Comment (LeonBrands) [inline legion/engine/core/ecs/registry.inl]: What's the future intent of this?

Comment (Ragingram2) [reply legion/engine/core/ecs/registry.inl]: This looks to be left over code. Initially we where going to register a serializer of component_type everytime we registered a new component type. We are now going a different direction using CPP attributes.

Comment (LeonBrands) [reply legion/engine/core/ecs/registry.inl]: Consider removing it to avoid confusion

File: legion/engine/core/ecs/registry.cpp [hunk 9]
```diff
@@ -60,11 +60,9 @@ namespace legion::core::ecs
     std::string Registry::getFamilyName(id_type id)
     {
         if (const auto itr = instance.m_familyNames.find(id); itr != instance.m_familyNames.end())
-        {
             return itr->second;
-        }
 
-        return "Component type " + std::to_string(id);
+        return "";
```
Comment (LeonBrands) [inline legion/engine/core/ecs/registry.cpp]: What is this used for and is the new empty return value properly handled?

File: legion/engine/core/ecs/registry.hpp [hunk 10]
```diff
@@ -205,6 +204,8 @@ namespace legion::core::ecs
         //template<typename component_type>
         //static component_type& createComponent(entity target, serialization::component_prototype<component_type>&& prototype);
 
+        static void* createComponent(id_type typeId, entity target, const void* component);
```
Comment (LeonBrands) [inline legion/engine/core/ecs/registry.hpp:207]: no documentation on this function and a commented out older version of the function above? Clean up commented out code if unused

File: legion/engine/core/filesystem/view.hpp [hunk 11]
```diff
@@ -144,6 +146,12 @@ namespace legion::core::filesystem
         mutable navigator::solution m_foundSolution{};
     };
 
+    class view_util
+    {
+    public:
```
Comment (LeonBrands) [inline legion/engine/core/filesystem/view.hpp]: this is created as a class but is named like a struct (view_util instead of ViewUtil). I'm also uncertain why this is created for a single function (get_view_path) which seemingly should just be a part of `view`?

File: legion/engine/core/serialization/serialization.hpp [hunk 12]
```diff
@@ -0,0 +1,33 @@
+#pragma once
```
Comment (LeonBrands) [inline legion/engine/core/serialization/serialization.hpp:1]: No documentation in this file at __all__

File: legion/engine/core/serialization/serialization.inl [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        PropagateErrors(result, warnings);
+
+        return { *data, warnings };
+    }
+
+    template<typename ViewType, typename Type>
+    inline common::result<Type, fs_error> load(const fs::view& file, std::string_view name)
+    {
+        ViewType view{};
+        byte rawData[sizeof(Type)];
```
Comment (LeonBrands) [inline legion/engine/core/serialization/serialization.inl:65]: Is it possible for this to get the wrong size if a `Type&` would be passed as template parameter? Perhaps a static assert or remove_reference_t should be in place. (see other uses of sizeof(Type) in this file as well)

File: legion/engine/core/serialization/serializationregistry.hpp [hunk 14]
```diff
@@ -1,27 +1,26 @@
...
 #include <core/containers/pointer.hpp>
+#include <core/serialization/serializer.hpp>
 
 #include <map>
-#include <any>
 
 namespace legion::core::serialization
 {
-    class Registry
+    class serializer_registry
```
Comment (LeonBrands) [inline legion/engine/core/serialization/serializationregistry.hpp]: if this was changed to snake case naming then it should also become a struct to follow conventions

File: legion/engine/core/filesystem/view.cpp [hunk 15]
```diff
@@ -422,4 +421,51 @@ namespace legion::core::filesystem
         //return empty ok
         return common::success;
     }
+
+
+    std::string view_util::get_view_path(const view& view, bool mustBeFile)
+    {
+
```
Comment (LeonBrands) [inline legion/engine/core/filesystem/view.cpp]: Excessive white-space, is this acceptable according to our standard?

File: legion/engine/core/types/meta.hpp [hunk 16]
```diff
@@ -223,4 +254,10 @@ namespace legion::core
 
     template<class T, typename... Args>
     inline constexpr bool is_brace_constructible_v = is_brace_constructible<T, Args...>::value;
+
+    template<typename T>
```
Comment (LeonBrands) [inline legion/engine/core/types/meta.hpp]: value could be `static constexpr bool`

**Review by LeonBrands — COMMENTED** [review]
I might've missed more, there's a lot of code in this PR which makes it very difficult to review it thoroughly. Try to make smaller PRs in the future as it may improve code quality.

File: legion/engine/core/ecs/registry.cpp (hunk 9 above)
Comment (GlynLeine) [reply legion/engine/core/ecs/registry.cpp]: yeah this was used, and then the empty string was used to check errors. but i'm pretty sure this can be reverted. other than that it is only used in imgui of our current makeshift editor. ultimately this fucntion should just return a result that can error out.

File: legion/engine/core/ecs/registry.hpp (hunk 10 above)
Comment (GlynLeine) [reply legion/engine/core/ecs/registry.hpp:207]: yeah i haven't touched anything outside of the serialization framework yet, since better ECS integration, scheduling integration, unit testing, and docs are planned for later subfeatures

File: legion/engine/core/filesystem/view.hpp (hunk 11 above)
Comment (GlynLeine) [reply legion/engine/core/filesystem/view.hpp]: hmmm i don't even know this class, i'll have a look at this

File: legion/engine/core/serialization/serialization.hpp (hunk 12 above)
Comment (GlynLeine) [reply legion/engine/core/serialization/serialization.hpp:1]: docs is for a later subfeature

**Comment by GlynLeine** [comment]
> I might've missed more, there's a lot of code in this PR which makes it very difficult to review it thoroughly. Try to make smaller PRs in the future as it may improve code quality.

yeah i know, this is a very old branch that snowballed. that's why i rly wanted to marge it and get rid of it so we could go back to smaller branches again

File: legion/engine/core/serialization/serialization.inl (hunk 13 above)
Comment (GlynLeine) [reply legion/engine/core/serialization/serialization.inl:65]: yeah static_assert would be good, passing pointer types or reference types would break compilation anyways. but at least with a static_assert we could tell you why.

File: legion/engine/core/serialization/serialization.hpp (hunk 12 above)
Comment (LeonBrands) [reply legion/engine/core/serialization/serialization.hpp:1]: seems odd but alright, since it's a subfeature that's acceptable I suppose

File: legion/engine/core/serialization/serializationregistry.hpp (hunk 14 above)
Comment (GlynLeine) [reply legion/engine/core/serialization/serializationregistry.hpp]: yeah this shouldn't be snake case

File: legion/engine/core/filesystem/view.cpp (hunk 15 above)
Comment (GlynLeine) [reply legion/engine/core/filesystem/view.cpp]: nope, but this class itself might need to be removed

File: legion/engine/core/types/meta.hpp (hunk 16 above)
Comment (GlynLeine) [reply legion/engine/core/types/meta.hpp]: it even should be, this makes no sense not to be

**Comment by GlynLeine** [comment]
@LeonBrands please not that this PR does not merge into develop

File: legion/engine/core/filesystem/view.cpp (hunk 15 above)
Comment (LeonBrands) [reply legion/engine/core/filesystem/view.cpp]: Either remove it or fix the whitespace before merge

**Comment by LeonBrands** [comment]
I realised - I'd still want to see the issues I mentioned get fixed - the less problems to fix on the develop merge in the future the better.

**Comment by GlynLeine** [comment]
> I realised - I'd still want to see the issues I mentioned get fixed - the less problems to fix on the develop merge in the future the better.

oh yeah i definetely agree! also ofc the build and analyze issues

File: legion/engine/core/types/meta.hpp [hunk 17]
```diff
@@ -254,10 +254,4 @@ namespace legion::core
 
     template<class T, typename... Args>
     inline constexpr bool is_brace_constructible_v = is_brace_constructible<T, Args...>::value;
-
-    template<typename T>
-    struct is_pointer { static const bool value = false; };
```
Comment (LeonBrands) [inline legion/engine/core/types/meta.hpp]: why were these removed instead of made constexpr?

Comment (GlynLeine) [reply legion/engine/core/types/meta.hpp]: nothing uses them, and the STL already has one. I don't know who made this or why they existed in the first place.

**Review by LeonBrands — DISMISSED** [review]
I think we're good but I can't guarantee that it's 100%. Will need a more thorough check on merge to develop
